### ISSUE TITLE
Replacing versus supplementing traversal of private uses and imports, continued

### ISSUE THREAD

**Issue by lydia-duncan** [chapel-lang/chapel#16357]
This is a continuation of #16183, intended to give folks a fresh entry point to the discussion so that we can finalize it.

Basically, we are trying to determine the set of rules used to determine whether a field or method can be resolved for an instance of a type that isn't necessarily visible to the current scope.  We've had various solutions for this problem over the years, but we always seem to run into a case where a user expects to be able to call a method on an instance they've obtained and can't.  Our latest solution causes us to always go into the scope where a type is defined to search for methods and fields on it.

The most recent previous solution involved ignoring the privacy and limitations of uses and imports when resolving method calls (and things like operators on the type).  The main drawback of this approach is that it means `private` and the other limitations are a lie w.r.t. methods and fields, which could be confusing to users.  The new solution handles many of the cases that motivated us to perform that previous traversal, so we wanted to determine whether this meant that we should stop performing that traversal as well or whether it was best to continue to do both.  In looking at the impact of no longer performing that traversal, only one test failure gave me pause.  The following test worked under the previous solution, but is not handled by going to the point where the type is defined (because we are trying to resolve a secondary method that was defined in another module entirely):

test/visibility/only/canSeeSecondaryMethod2.chpl:
```chapel
module OuterModule {
  use secondaryMethod only Foo; // The module secondaryMethod defines type `Foo`
  use Extender only Foo;
  // Verifies that you can define a method in a module outside of the original
  // type definition, and by specifying that type in your 'only' list, access
  // it.

  var a = new owned Foo(7);
  writeln(a.innerMethod(3)); // Should be 21
  writeln(a.newMethod(1)); // Should be 8

  module Extender {
    use secondaryMethod only Foo;

    proc Foo.newMethod(z: int) {
      return field + z;
    }

    proc unrelated() {
      writeln("I'm unrelated to Foo!");
    }
  }
}
```

(I've included the helper file and output in the following details section)

<details>

Here's the definition of secondaryMethod.chpl
```chapel
class Foo {
  var field = 14;

  proc innerMethod(x: int) {
    return field * x;
  }
}

proc Foo.outerMethod(y: int) {
  return field - y;
}

proc other() {
  writeln("I don't involve Foo at all!");
}

proc returnAFoo(x: int) {
  return new owned Foo(x);
}
```

The failure output is:
```
canSeeSecondaryMethod2.chpl:10: error: unresolved call 'owned Foo.newMethod(1)'
canSeeSecondaryMethod2.chpl:10: note: because no functions named newMethod found in scope
```

</details>

It seems like something that would cause a user frustration if it did not work.

Related side bar (included for completeness but not necessary for understanding):
<details>
In the other issue, Vass brought up that he'd like listing `Foo` to only mention the type `Foo` and to have the method names be listable in the limitation clause.  This is currently not on the table but I wanted to mention it in case others had the same idea.  It's partially due to uncertainty about handling things like operators or generated methods the user doesn't have to know about but the compiler requires, and partially due to trickiness with methods being defined within the scope of the type itself and so not being as easily found when we verify the symbols in a limitation clause are actually accessible.
</details>

Even more alternatively - prior to the previous solution, we would track symbols associated with a type, so that when the type was mentioned in an except or only list, we would allow or block its methods appropriately.  Now obviously that wasn't sufficient on its own (or we wouldn't have had to consider the more recent changes).  However, we could choose to restore that solution while also stopping the traversal of uses that I described earlier, to ensure the above test still continues to pass.

Ideally, I'd like us to determine a path forward that minimizes the possibility of users asking for a method to be available that wouldn't be available otherwise.  While I would prefer to stop ignoring aspects of use and import statements in order to find these methods, I'd much prefer not having to alter the traversal strategy again after this change because we forgot another pattern.

So, in summary, our options as I see them are:
A. Continue to support both going to the type's definition point and traversing any uses and imports we find, regardless of their privacy or limitations
B. Go to the type's definition point, but return to respecting the privacy and limitations of uses and imports, accepting that this test will fail (and that a user will probably try to argue with us about it)
C. Go to the type's definition point, return to respecting the privacy and limitations of uses and imports, and restore tracking of symbols associated with a type name (and hope that is enough that users won't find other edge cases we haven't thought of)

If there's another option you can think of, please let me know.

**Comment by lydia-duncan** [comment]
Michael and I talked about this offline Wednesday.  He views this issue as centered around whether secondary methods should be brought into scope when the type on which they are defined is explicitly brought in to scope.  In his opinion, this should not be the case but he would like there to be a way to opt in to doing so.  One potential strategy is to enable it via re-exporting the type (I made some tests of some options, but none of them struck me as obviously the right path and there was plenty of variation between their behavior today, so I don't think I'll commit them just yet).  But he views the discussion of how to enable it as kind of separate from whether secondary methods should be brought in via this particular strategy, though obviously related.

A couple of items came up in our discussion:
- what happens when an operator is labeled as private?  There didn't appear to be tests of this before, but it seems to work the way I expected, so I'll commit those tests.
- Michael had a memory of Rust doing something similar about ways to bring methods into scope.  I'm not sure if [this](https://doc.rust-lang.org/reference/visibility-and-privacy.html#pubin-path-pubcrate-pubsuper-and-pubself) is what he's referring to - it doesn't seem to be a 1:1 correspondence of concepts, but looking at [this](https://doc.rust-lang.org/reference/items/associated-items.html#associated-functions-and-methods) doesn't give me a better candidate for what he's thinking of.  I'd be curious if other people that are more familiar with Rust have more insight.

**Comment by vasslitvinov** [comment]
To me `canSeeSecondaryMethod2.chpl` should not compile. Because `use Extender only Foo` tells me that I am using only the name `Foo`. It explicitly tells me not to use any other names, such as `newMethod`.

So let us go with "our latest solution" and drop "the most recent previous solution" aka "performing that traversal".

Some features we could add now or in the future:

(A) To include/exclude **_all_** methods on the type Foo, write `use M only Foo.*`

(B) To include/exclude a given method on the type Foo, write `use M only Foo.methodName`

(C+) We can define `use M only someName` to include methods named _someName_, on any type, in addition to functions and types named _someName_.

(C-) Or, we can define `use M only someName` not to include methods _someName_, on any type. The user must write `only Foo.someName` or perhaps even `only *.someName` to include methods _someName_ on _Foo_ or on any type. In this case, `except someName` does not exclude methods _someName_.

(D) When a call resolves to a function returning a type _Foo_, the caller is allowed to invoke any methods on _Foo_ that are visible from the callee.

**Comment by bradcray** [comment]
> To me canSeeSecondaryMethod2.chpl should not compile. Because use Extender only Foo tells me that I am using only the name Foo. It explicitly tells me not to use any other names, such as newMethod.

At a sniff, I'm not sure I agree with this.  I think of methods and fields on a record or object as being names that are scoped by that type name, similar to how symbols within a module are names that are scoped by that module's name.  If I say `use MyModule only MySubModule;` I can access all the symbols within `MySubmodule` via `MySubmodule.symbol`.  My intuition is that `only Foo` should be similar for a class or record `Foo`.  Here's my example:

```chapel
module M {
  module N {
    var foo = 42;
  }
}

module O {
  proc main() {
    use M only N;
    writeln(N.foo);
  }
}
```

Granted, records/classes aren't the same as modules, but this is my intuition for why having access to a type's symbol should give me access to methods and fields on that symbol as well (even for a secondary method).  Or at least, I'm not seeing the argument for why they should be treated differently.

At the meta-level on the desire to get more opinions on this question:  I'd still love to see this issue turned into one that's strictly about the language definition and behavior of an example program rather than the compiler's implementation and behavior so that it would be an easy one to point end-users to who don't know or care how the compiler works.  I'm hesitant to point users to this issue because it doesn't feel sufficiently focused on the language and example (e.g., most of the key details are hidden away while most of the top-level text is about the compiler's behavior).

**Comment by vasslitvinov** [comment]
> I think of methods and fields on a record or object as being names that are scoped by that type name, similar to how symbols within a module are names that are scoped by that module's name. If I say `use MyModule only MySubModule`; I can access all the symbols within MySubmodule via MySubmodule.symbol. My intuition is that `only Foo` should be similar for a class or record Foo.

This sounds reasonable to me.

Given this adjustment, visibility rules for methods and fields look like this:

* #include visibility rules for non-methods, perhaps with fine-tuning for `except`/`only methodName`
* public methods and fields on the receiver that are visible from the receiver type's declaration point
  are always visible
* `except`/`only typeName` also affect methods and fields of the type `typeName`
* (future extension) `except`/`only typeName.name` affect typeName's methods and fields called `name`
* (future extension) `use typeName` by analogy to `use ModuleName` makes typeName's type methods available directly

**Comment by lydia-duncan** [comment]
> (D) When a call resolves to a function returning a type Foo, the caller is allowed to invoke any methods on Foo that are visible from the callee.

This is arguably the world we live in today.

I agree with Brad's comment, but I am relatively in favor of A) and B), and I think as a result am completely on board with Vass's latest comment.

> At the meta-level on the desire to get more opinions on this question: I'd still love to see this issue turned into one that's strictly about the language definition and behavior of an example program rather than the compiler's implementation and behavior so that it would be an easy one to point end-users to who don't know or care how the compiler works. I'm hesitant to point users to this issue because it doesn't feel sufficiently focused on the language and example (e.g., most of the key details are hidden away while most of the top-level text is about the compiler's behavior).

I should point out that I don't actually mention the compiler itself in the issue description.  While this is the compiler's behavior, the various strategies described are the **policies** being followed by the implementation.  And while the motivation behind why we decided to follow each of these strategies is because the other strategies did not cover all the cases we wanted (making the implementation ultimately insufficient), the policies are really what I think should be discussed.  When a user asks "why is/isn't this method being found here", you'll answer them using the policy being employed.

**Comment by mppf** [comment]
> > To me canSeeSecondaryMethod2.chpl should not compile. Because use Extender only Foo tells me that I am using only the name Foo. It explicitly tells me not to use any other names, such as newMethod.
> 

@bradcray  wrote

> At a sniff, I'm not sure I agree with this. 

I just want to make sure that it's clear that we're talking about secondary methods defined in a different module from the type. We already can access secondary methods defined (or, presumably, re-exported) by the module defining the type.

There isn't an analogous case where one can add a variable to another module.

---

First, let's start with an example of a secondary method defined in the module defining the type:

``` chapel
module DefinesFoo {
  record Foo { }
  proc Foo.frobnify() { }
}

module ReturnsFoo {
  proc makeFoo() {
    use DefinesFoo;
    return new Foo();
  }
}

module UserOfFoo {
  import ReturnsFoo;
  var x = ReturnsFoo.makeFoo();
  x.frobnify(); // this should compile
}
```

I expect that this first example will compile just fine after PR #16409 . I don't think this issue is debating that in any way.

---

Now, let's consider an example making the secondary method in a different module. As far as I know, the behavior of this kind of program is the only thing being debated here (other than the compiler implementation).

``` chapel
module DefinesBar {
  record Bar { }
}

module DefinesBarMethod {
  use DefinesBar;
  proc Bar.doSomething() { }
}

module ReturnsBar {
  proc makeBar() {
    use DefinesBar;
    return new Bar();
  }
}

module UserOfBar1 {
  var x = makeBar();
  x.doSomething(); // compiler error - doSomething is not found
                   // Further, I don't think anybody is (yet) proposing that this should compile.
                   // It is not the subject of this issue.
}

module UserOfBar2 {
  use DefinesBarMethod only ;
  var x = makeBar();
  x.doSomething(); // Should this compile?
                   // This is what we are debating in this issue.
                   // Should the fact that DefinesBarMethod was use/only'd or private use'd
                   // cause Bar.doSomething() to be visible?
}
```

In this issue, we are deciding if `UserOfBar2` should compile or not. If we want to debate whether `UserOfBar1` should compile, we should do that in a new issue.

It is my opinion that neither `UserOfBar1` nor `UserOfBar2` should compile.

**Comment by gbtitus** [comment]
> I expect that this first example will compile just fine after PR #16409 . I don't think this issue is debating that in any way.

Are you sure that's the PR you meant to reference?  It looks like there might have been a cut/paste error there ... :-)

**Comment by bradcray** [comment]
> I just want to make sure that it's clear that we're talking about secondary methods defined in a different module from the type.

That's clear to me, so I must have been unclear.

In terms of the example you just added, what I was trying to say was that within `UserOfBar2`, my intuition is that it'd be sufficient for me to say: 

```chapel
use DefinesBarMethod only Bar;
```

in order to access `Bar.doSomething();` and that I should not need to (or be able to) say:

```chapel
use DefinesBarMethod only doSomething;
``

(because `doSomething` isn't scoped by `DefinesBarMethod`).  It also doesn't seem like I should be required to say:

```chapel
use DefinesBarMethod only Bar.doSomething;
```

which is what I thought Vass was suggesting when he wrote:

> Because `use Extender only Foo` tells me that I am using only the name `Foo`. It explicitly tells me not to use any other names, such as newMethod.

The reason I say this is because it seems like overkill given that I wouldn't need to write `use DefinesBar only Bar.primaryMethod, Bar.secondaryMethod;` in order to use methods defined directly within `DefinesBar`.

All that said, I'm not at all confident that I'm right about this, it was just my quick reaction to Vass's statement.  Maybe getting access to secondary methods within a module that doesn't define a type should require more work for some reason that's not occurring to me.  Or maybe I shouldn't even be able to get at secondary methods within a module unless the module re-exports the type itself.  In your example, maybe it's only legal for me to write `use DefineBarMethod only Bar;` if and when `DefinesBarMethod` contained  `public use DefinesBar;`?

In your example, you wrote:

> `// This is what we are debating in this issue.`

but I thought the equivalent of the OP for your example would've been to say `use DefinesBar only Bar;` not `... only ;` (though your question is obviously interesting too).

This example is closer to what I was looking for to put before users as a language design question.  Though I might drop the second `use` statement from `UserOfBar2` and ask what `use` statement(s) might I write here in order to call `x.doSomething()`, where choices might be "none", `use DefinesBarMethod;`, `use DefinesBarMethod only ;`, or `use DefinesBarMethod only [something here?];`.

**Comment by vasslitvinov** [comment]
Summarizing the above discussion in #16415.

**Comment by mppf** [comment]
It sounds like we are agreeing that my `UserOfBar2` should not compile as written. That might be enough of an answer to the original question.

> 
> In your example, you wrote:
> 
> > `// This is what we are debating in this issue.`
> 
> but I thought the equivalent of the OP for your example would've been to say `use DefinesBar only Bar;` not `... only ;` (though your question is obviously interesting too).

This goes back to something in the original post:

> ... solution involved ignoring the privacy and limitations of uses and imports when resolving method calls

If we are ignoring privacy and limitations of uses when resolving method calls, both `  use DefinesBarMethod only ;` as well as `  use DefinesBarMethod only Bar;` would find `doSomething`.

(We also don't currently have an example showing how a private use of say `DefinesOtherSecondaryMethod` inside of `DefinesBarMethod` would allow that secondary method to be available from a `use DefinesBarMethod only;`. I personally think that this is similar to `UserOfBar2` and should not compile.)

> 
> In terms of the example you just added, what I was trying to say was that within `UserOfBar2`, my intuition is that it'd be sufficient for me to say: 
> 
> ```chapel
> use DefinesBarMethod only Bar;
> ```
> 

That would be fine with me and is something I think we can discuss more in #16415.

**Comment by bradcray** [comment]
> It sounds like we are agreeing that my UserOfBar2 should not compile as written.

I didn't mean to imply that.  I'm not sure what I think of your example as written and was trying to avoid taking a position on it in my response yesterday.

**Comment by lydia-duncan** [comment]
Closing this issue, since we discussed it further in a group deep dive and ultimately decided to return to respecting the privacy and limitations of use and import statements when resolving methods.
